File: 2018/day05_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any


def solution(content: str) -> Any:
    return min(
        len(react(remove(unit, content))) for unit in {ch.lower() for ch in content}
    )


def remove(unit: str, polymer: str):
    return "".join(ch for ch in polymer if ch != unit and ch.lower() != unit)


def react(polymer: str):
    copy = polymer[:]
    is_on = True
    while is_on:
        is_on = False
        for index, (fst, snd) in enumerate(zip(copy, copy[1:])):
            if should_react(fst, snd):
                is_on = True
                copy = copy[:index] + copy[index + 2 :]
                break
    return copy


def should_react(a: str, b: str):
    return a.lower() == b.lower() and a != b


def main() -> None:
    with open("day05.dat", "r", encoding="utf8") as file:
        print(solution(file.read().strip()))


if __name__ == "__main__":
    main()

File: 2018/day05-2.ts
function solution(content: string) {
  return Math.min(
    ...Array.from(new Set(content.toLowerCase()))
      .map((ch) => content.split("").filter((it) => it.toLowerCase() !== ch))
      .map(react)
      .map((result) => result.length)
  );
}

function react(polymer: string[]) {
  const copy = polymer.slice();
  let isOn = true;

  while (isOn) {
    isOn = false;
    for (let index = 0; index < copy.length; index++) {
      if (
        index < copy.length - 1 &&
        copy[index].toLowerCase() === copy[index + 1].toLowerCase() &&
        copy[index] !== copy[index + 1]
      ) {
        isOn = true;
        copy.splice(index, 2);
        break;
      }
    }
  }

  return copy;
}

Deno.readTextFile("day05.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
